site/src/contexts/ProxyContext.tsx

```diff
@@ -0,0 +1,211 @@
+        }}
+      >
+        {children}
+      </ProxyContext.Provider>
+    )
+  }
+  // ******************************* //
+
+  // TODO: @emyrk Should make an api call to /regions endpoint to update the
+  // regions list.
+
+  return (
+    <ProxyContext.Provider
+      value={{
+        proxy: proxy,
+        isLoading: regionsLoading,
+        error: regionsError,
+        // A function that takes the new regions and selected region and updates
+        // the state with the appropriate urls.
+        setProxy: setAndSaveProxy,
+      }}
+    >
+      {children}
+    </ProxyContext.Provider>
+  )
+}
+
+export const useProxy = (): ProxyContextValue => {
+  const context = useContext(ProxyContext)
+
+  if (!context) {
+    throw new Error("useProxy should be used inside of <ProxyProvider />")
+  }
+
+  return context
+}
+
+/**
+ * getURLs is a helper function to calculate the urls to use for a given proxy configuration. By default, it is
+ * assumed no proxy is configured and relative paths should be used.
+ * Exported for testing.
+ *
+ * @param regions Is the list of regions returned by coderd. If this is empty, default behavior is used.
+ * @param selectedRegion Is the region the user has selected. If this is undefined, default behavior is used.
+ */
+export const getURLs = (
+  regions: Region[],
+  selectedRegion?: Region,
+): PreferredProxy => {
+  // By default we set the path app to relative and disable wildcard hostnames.
+  // We will set these values if we find a proxy we can use that supports them.
+  let pathAppURL = ""
+  let wildcardHostname = ""
+
+  // If a region is selected, make sure it is in the list of regions. If it is not
+  // we should default to the primary.
+  selectedRegion = regions.find(
+    (region) => selectedRegion && region.id === selectedRegion.id,
+  )
+
+  if (!selectedRegion) {
+    // If no region is selected, default to the primary region.
+    selectedRegion = regions.find((region) => region.name === "primary")
+  }
+
+  // Only use healthy regions.
+  if (selectedRegion && selectedRegion.healthy) {
+    // By default use relative links for the primary region.
+    // This is the default, and we should not change it.
+    if (selectedRegion.name !== "primary") {
+      pathAppURL = selectedRegion.path_app_url
+    }
+    wildcardHostname = selectedRegion.wildcard_hostname
+  }
+
+  // TODO: @emyrk Should we notify the user if they had an unhealthy region selected?
+
+  return {
+    selectedRegion,
+    // Trim trailing slashes to be consistent
+    preferredPathAppURL: pathAppURL.replace(/\/$/, ""),
+    preferredWildcardHostname: wildcardHostname,
+  }
+}
+
+// Local storage functions
+
+export const savePreferredProxy = (saved: PreferredProxy): void => {
+  window.localStorage.setItem("preferred-proxy", JSON.stringify(saved))
+}
+
+export const loadPreferredProxy = (): PreferredProxy | undefined => {
+  const str = localStorage.getItem("preferred-proxy")
+  if (str === undefined || str === null) {
```

Comment:
[It looks to me](https://developer.mozilla.org/en-US/docs/Web/API/Storage/getItem) like this is either going to return a value, or else it's going to return `null`, so I think we can simplify:

```suggestion
  if (!str) {
```